cmake/OpenCVDownload.cmake

```diff
@@ -0,0 +1,124 @@
+#
+#  Download and optionally unpack a file
+#
+#  ocv_download(FILENAME p HASH h URL u DESTINATION_DIR d [STATUS s] [UNPACK] [RELATIVE])
+#    FILENAME - filename
+#    HASH - MD5 hash
+#    URL - full download url
+#    DESTINATION_DIR - file will be copied to this directory
+#    STATUS - passed variable will be updated in parent scope,
+#             function will not fail the build in case of download problem if this option is provided,
+#             but will fail in case when other operations (copy, remove, etc.) failed
+#    UNPACK - downloaded file will be unpacked to DESTINATION_DIR
+#    RELATIVE - if set, then URL is treated as base, and FILENAME will be appended to it
+#  Note: uses OPENCV_DOWNLOAD_DIR folder as cache, default is <opencv>/.cache
+
+#
+# Main function
+#
+
+set(OPENCV_DOWNLOAD_DIR "${OpenCV_SOURCE_DIR}/.cache" CACHE PATH "Download cache dir")
+set(OPENCV_DOWNLOAD_LOG "${OpenCV_BINARY_DIR}/CMakeDownloadLog.txt")
+
+# Init downlod cache directory and log file
+if(NOT EXISTS "${OPENCV_DOWNLOAD_DIR}")
+  file(MAKE_DIRECTORY ${OPENCV_DOWNLOAD_DIR})
+endif()
+file(WRITE "${OPENCV_DOWNLOAD_LOG}" "use_cache \"${OPENCV_DOWNLOAD_DIR}\"\n")
+
+
+function(ocv_download)
+  cmake_parse_arguments(DL "UNPACK;RELATIVE_URL" "FILENAME;HASH;URL;DESTINATION_DIR;STATUS" "" ${ARGN})
+
+  ocv_assert(DEFINED DL_FILENAME)
+  ocv_assert(DEFINED DL_HASH)
+  ocv_assert(DEFINED DL_URL)
+  ocv_assert(DEFINED DL_DESTINATION_DIR)
+
+  set(${DL_STATUS} TRUE PARENT_SCOPE)
+
+  if(DL_RELATIVE_URL)
+    set(DL_URL "${DL_URL}${DL_FILENAME}")
+  endif()
+
+  set(mode "copy")
+  if(DL_UNPACK)
+    set(mode "unpack")
+  endif()
+
+  # Log all calls to file
+  file(APPEND "${OPENCV_DOWNLOAD_LOG}" "do_${mode} \"${DL_FILENAME}\" \"${DL_HASH}\" \"${DL_URL}\" \"${DL_DESTINATION_DIR}\"\n")
+
+  # ... and to console
+  message(STATUS "Download (${mode}):")
+  message(STATUS "  ${DL_FILENAME} (${DL_HASH})")
+  message(STATUS "  ${DL_URL}")
+  message(STATUS "  ${DL_DESTINATION_DIR}")
+
+  # Copy mode: check if copy destination exists and is correct
+  if(NOT DL_UNPACK)
+    set(COPY_DESTINATION "${DL_DESTINATION_DIR}/${DL_FILENAME}")
+    if(EXISTS "${COPY_DESTINATION}")
+      file(MD5 "${COPY_DESTINATION}" target_md5)
```

Comment:
This check is for _"copy single file"_ mode only, so unpacks are not affected. And most of such files will be copied to the source folder (descriptors in xfeatures2d). 
I think it is better to leave it as is because it is more simple, less error prone and doesn't affect performance too much.